client/mksubvolume.cc

```diff
@@ -0,0 +1,420 @@
+/*
+ * Copyright (c) [2015] SUSE LLC
+ *
+ * All Rights Reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of version 2 of the GNU General Public License as published
+ * by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact Novell, Inc.
+ *
+ * To contact Novell about this file by physical or electronic mail, you may
+ * find current contact information at www.novell.com.
+ */
+
+
+#include <string.h>
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <ext2fs/ext2_fs.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+
+#include <iostream>
+#include <boost/algorithm/string.hpp>
+
+#include "snapper/BtrfsUtils.h"
+#include "snapper/AppUtil.h"
+#include "snapper/MntTable.h"
+#include "utils/GetOpts.h"
+
+
+using namespace snapper;
+using namespace BtrfsUtils;
+
+
+string target;
+
+bool set_nocow = false;
+bool verbose = false;
+
+
+void
+do_tmp_mount(const libmnt_fs* fs, const char* tmp_mountpoint, const string& subvol_option)
+{
+    if (verbose)
+	cout << "do-tmp-mount" << endl;
+
+    libmnt_fs* x = mnt_copy_fs(NULL, fs);
+    if (!x)
+	throw runtime_error("mnt_copy_fs failed");
+
+    struct libmnt_context* cxt = mnt_new_context();
+    mnt_context_set_fs(cxt, x);
+    mnt_context_set_target(cxt, tmp_mountpoint);
+    if (!subvol_option.empty())
+	mnt_context_set_options(cxt, ("subvol=" + subvol_option).c_str());
+    else
+	mnt_context_set_options(cxt, "subvolid=5");
+
+    int ret = mnt_context_mount(cxt);
+    if (ret != 0)
+	throw runtime_error(sformat("mnt_context_mount failed, ret:%d", ret));
+
+    mnt_free_context(cxt);
+    mnt_free_fs(x);
+}
+
+
+void
+do_create_subvolume(const string& tmp_mountpoint, const string& subvolume_name)
+{
+    if (verbose)
+	cout << "do-create-subvolume" << endl;
+
+    string parent = tmp_mountpoint + "/" + dirname(subvolume_name);
+
+    if (mkdir(parent.c_str(), 0777) != 0 && errno != EEXIST)
+	throw runtime_error_with_errno("mkdir failed", errno);
```

Comment:
The program does not setup logging functions so ST_THROW does not bring benefits.